### ISSUE TITLE
I registed notification for FCM but I get the message from GCM service.

### ISSUE THREAD

**Issue by jessejiang0214** [Azure/azure-notificationhubs-android#6]
Hi there,

We use notification template to send notification in the backend, here is the code
```
    public class HubNotifier : IHubNotifier
    {
        protected NotificationHubClient NotificationHubClient { get; }
        public HubNotifier(string connectionString, string hubId)
        {
            this.NotificationHubClient =
                NotificationHubClient.CreateClientFromConnectionString(
                    connectionString,
                    hubId);
        }
        public async Task<bool> SendNotificationsAsync(string title, string body, string notificationId, string cardId, string userTag)
        {
            IDictionary<string, string> properties = new Dictionary<string, string>()
                                                         {
                                                             { "title", title },
                                                             { "body", body },
                                                             { "notificationId", notificationId },
                                                             { "cardId", cardId }
                                                         };
            var template = new TemplateNotification(properties) { Body = "{\"title\":\"${title}\",\"body\":\"${body}\",\"notificationId\":\"${notificaitonId}\",\"cardId\":\"${cardId}\"}" };
            var result = await this.NotificationHubClient.SendNotificationAsync(template, userTag);
            return result.Success == 0;
        }
        public async Task<CollectionQueryResult<RegistrationDescription>> GetRegistrationDescriptions()
        {
           return await this.NotificationHubClient.GetAllRegistrationsAsync(100);
        }
    }
```

For Android client, I called RegisterTemplate to register the template 
```
                    _hub.RegisterTemplate(FirebaseInstanceId.Instance.Token,
                                                      NotificationHelper.TemplateName,
                                                      NotificationHelper.GetNotificationTemplate(),
                                                      tags.ToArray());
```
And follow this document to do the message receice 
[Get started with Notification Hubs for Xamarin.Android apps](https://docs.microsoft.com/en-us/azure/notification-hubs/xamarin-notification-hubs-push-notifications-android-gcm)

Everything is fine, I can send the notification from the portal just like the document said, I select the platform as Android it works fine.

But, our backend need to send in a template, so the bug is if I select the template to send the notification, the Android devices will get the notification, but it won't call OnMessageReceived in FirebaseMessagingService
```
    [Service]
    [IntentFilter(new[] { "com.google.firebase.MESSAGING_EVENT" })]
    public class MyFirebaseMessagingService : FirebaseMessagingService
    {        
        public override void OnMessageReceived(RemoteMessage message)
        {
            if (message.GetNotification() != null)
            {
                SendNotification(message.GetNotification().Body);
            }
            else
            {
                //Only used for debugging payloads sent from the Azure portal
                SendNotification(message.Data.Values.First());

            }

        }
```

## root cause:
I have a feeling that in the backend, the template didn't send to FCM, maybe GCM, so I implement the GCM client in Android device, and it gets the message
```
    [Service (Exported = false), IntentFilter (new [] { "com.google.android.c2dm.intent.RECEIVE" })]
    public class MyGcmListenerService: GcmListenerService
    {
        public override void OnMessageReceived(string from, Bundle data)
        {
            // Extract the message received from GCM:
            var message = data.GetString("message");

            System.Diagnostics.Debug.WriteLine($"Get GCM notification {message}");
            // Forward the received message in a local notification:
            SendNotification(message);
        }
```

And, there is no document to tell developers to how to use the notification template on server and clinet.

**Comment by boindiil** [comment]
This issue lead to even more problems if you are using _SendDirectNotificationAsync_ in combination with _FcmNotification_. If you are using this combination you are not able to send notifications to Android devices, although sending via test-send in Azure portal works. I wasted about 4 hours trying to get it working.

In the end I just reverted to Version 2.0.2 using _GcmRegistration_s and _GcmNotification_s which works in combination with a registered FCM-Service in the notification hub.

**Comment by brannon** [comment]
@boindiil I asked someone to take a look at the issue you reported. Can you provide any more information or a snippet of sample code?

**Comment by RawsomeGH** [comment]
@brannon I'm experiencing the same problem, I think.

If I run `var registrationDescription = await notificationHubClient.GetRegistrationAsync<RegistrationDescription>(regId);` the type of **registrationDescription** is **GcmRegistrationDescription** even though I am expecting an **FcmRegistrationDescription**.

This means that I can't do anything at all with it since that type doesn't even exist in the current version ([see the docs](https://docs.microsoft.com/en-us/dotnet/api/microsoft.azure.notificationhubs.registrationdescription?view=azure-dotnet) which don't contain GcmRegistrationDescription as Derived either).

**Comment by boindiil** [comment]
@brannon with my second try I got it working, so I don't think there is an issue with sending native notifications.

**Comment by brannon** [comment]
@boindiil that's good to hear. We've made a few fixes in the SDK to work around some issues between GCM and FCM.

@RawsomeGH I think you are seeing a different problem. I'll ask someone to take a look.

**Comment by marstr** [comment]
Looks to the naive eye like this should have been marked fixed and closed when #46 was merged. I'm going to close this issue. Please let me know if this issue wasn't resolved, or open a new issue.
